File: 2019/rust/01.rs
// Advent of Code 2019, day 1
// (c) aichingert

fn part1(inp: &[u32]) -> u32 {
    inp.iter().map(|n| n / 3 - 2).sum::<u32>()
}

fn part2(inp: &mut [u32]) -> u32 {
    let mut ans = 0;

    inp.iter_mut().for_each(|n| 
        while *n / 3 > 1 {
            *n = *n / 3 - 2;
            ans += *n;
        }
    );

    ans
}

fn main() {
    let mut inp = std::fs::read_to_string("../input/01").unwrap().lines().map(|n| n.parse::<u32>().unwrap()).collect::<Vec<u32>>();

    println!("Part 1: {}", part1(&inp));
    println!("Part 2: {}", part2(&mut inp));
}

File: 2017/julia/01.jl
# Advent of Code 2017, day 1
# (c) aichingert

module Day1

function part1(s::String)
	ans = if s[begin] == s[end] Int32(s[begin]) - 48
	else 0 end

	for i in 1:length(s)-1
		if s[i] == s[i+1]
			ans += Int32(s[i]) - 48
		end
	end

	return ans
end

function part2(s::String)
	ans = 0
	half = length(s)÷2-1

	for i in eachindex(s)
		if s[i] == s[(half + i) % length(s)+1] 
			ans += Int32(s[i]) - 48
		end
	end

	return ans
end

input = readline("../input/01")

println("Part 1: $(part1(input))")
println("Part 2: $(part2(input))")

end
